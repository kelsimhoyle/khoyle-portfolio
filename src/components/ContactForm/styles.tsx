import styled from "styled-components";
import { colors } from "../../styles/utils";


export const Form = styled.form`
    width: 40vw;
    paddng: 10vw;
    margin: 5vw;
    display: flex;
    flex-direction: column;
    background-color: white;

    label {
        margin: .5rem;
        padding: .5rem;
    }
    
    input,
    textarea {
        width: 80%;
        padding: 5px;
        box-sizing: border-box;
        border-radius: 5px;
    }

    textarea {
        resize: none;
        height: 150px;
    }


    p {
        margin: 0;
    }

    button {
        color: white;
        background-color: ${colors.secondary};
        font-size: 1.5rem;
        border: none;
        padding: .5rem;
        cursor: pointer;
        text-decoration: none
    }

`

export const Btn = styled.button`

`