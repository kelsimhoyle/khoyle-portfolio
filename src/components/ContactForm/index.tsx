import React, { useState } from "react";
import addContact from "../../functions/addContact";
import { Form } from "./styles";

export type ContactDataType = {
    Name: String;
    Email: String;
    Message: String;
}

const ContactForm: React.FC<{}> = () => {
    const [data, setData] = useState({ errors: [], submitting: false, submitted: false });
    const [contactData, setContactData] = useState<ContactDataType>({ Name: "", Email: "", Message: "" });

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => setContactData({
        ...contactData,
        [e.currentTarget.name]: e.currentTarget.value
    });

    const handleSubmit = (e: React.SyntheticEvent) => {
        e.preventDefault();
        setData({ ...data, submitting: true });
        const { Name, Email, Message } = contactData;

        if (addContact({ Name, Email, Message })) return setData({ ...data, submitting: false, submitted: true });

        return setData({ ...data, submitting: false, submitted: false, errors: [...data.errors, "Error submitting contact request."] })
    }

    return (
        <>
            <Form onSubmit={handleSubmit} data-testid="form" name="contact-form">
                <label htmlFor="Name">
                    Name
                </label>

                <input
                    type="text"
                    name="Name"
                    value={contactData.Name as string}
                    onChange={handleInputChange}
                    placeholder="Name"
                    data-testid="name-input"
                />
                <label htmlFor="Email">
                    E-Mail
                </label>

                <input
                    type="text"
                    name="Email"
                    value={contactData.Email as string}
                    onChange={handleInputChange}
                    placeholder="Email"
                    data-testid="email-input"
                />
                <label htmlFor="Message">
                    Message
                </label>

                <textarea
                    rows={5}
                    name="Message"
                    value={contactData.Message as string}
                    onChange={handleInputChange}
                    placeholder="Message"
                    data-testid="message-input"
                />
                <button
                    type="submit"
                    data-testid="button"
                >Submit</button>
            </Form>

            {
                data.errors.length > 0 ? (
                    data.errors.map(error => <p>{error}</p>)
                ) : null
            }

            {data.submitting ? <p>Submitting contact request...</p> : null}

            {data.submitted ? <p className="submitted">Thank you! I will get back to you within 2 business days!</p> : null}
        </>
    );
};

export default ContactForm