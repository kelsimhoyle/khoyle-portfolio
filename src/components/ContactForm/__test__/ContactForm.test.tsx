import React from "react";
import { render, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer";

import ContactForm from "..";

let getByTestId;
let getByText;


const onSubmit = jest.fn(e => e.preventDefault())

beforeEach(() => {
    jest.clearAllMocks();
    const component = render(<ContactForm />);
    getByTestId = component.getByTestId;
    getByText = component.getByText;
});

test("contact buttons renders correctly", () => {
    const tree = renderer.create(<ContactForm />).toJSON()
    expect(tree).toMatchSnapshot()
});

test("input values change", () => {
    const nameInput = getByTestId("name-input");
    const emailInput = getByTestId("email-input");
    const messageInput = getByTestId("message-input");

    fireEvent.change(nameInput, {
        target: {
            value: "Kelsi"
        }
    });

    expect(nameInput.value).toBe("Kelsi");

    fireEvent.change(emailInput, {
        target: {
            value: "kelsi@kelsihoyle.com"
        }
    });

    expect(emailInput.value).toBe("kelsi@kelsihoyle.com");

    fireEvent.change(messageInput, {
        target: {
            value: "Hello!"
        }
    });

    expect(messageInput.value).toBe("Hello!");

});

test("validation works", () => {

});

test("form submits", () => {
    const form = getByTestId("form");
    const nameInput = getByTestId("name-input");
    const emailInput = getByTestId("email-input");
    const messageInput = getByTestId("message-input");

    const expected = {
        Name: "Kelsi",
        Email: "kelsi@kelsihoyle.com",
        Message: "Hello!"
    };

    fireEvent.change(nameInput, {
        target: {
            value: "Kelsi"
        }
    });

    fireEvent.change(emailInput, {
        target: {
            value: "kelsi@kelsihoyle.com"
        }
    });

    fireEvent.change(messageInput, {
        target: {
            value: "Hello!"
        }
    });

    fireEvent.submit(form);

    

   const submittedDiv = getByText("Thank you! I will get back to you within 2 business days!");

   expect(submittedDiv).toBeInTheDocument();
});


test("button submits form", () => {
    const btn = getByTestId("button");

    fireEvent.click(btn);

    const submittedDiv = getByText("Thank you! I will get back to you within 2 business days!");

    expect(submittedDiv).toBeInTheDocument();
});





