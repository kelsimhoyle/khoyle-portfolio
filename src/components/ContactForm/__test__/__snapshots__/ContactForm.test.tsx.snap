// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact buttons renders correctly 1`] = `
<form
  className="sc-bdfBQB cdNhYF"
  data-testid="form"
  name="contact-form"
  onSubmit={[Function]}
>
  <label
    htmlFor="Name"
  >
    Name
  </label>
  <input
    data-testid="name-input"
    name="Name"
    onChange={[Function]}
    placeholder="Name"
    type="text"
    value=""
  />
  <label
    htmlFor="Email"
  >
    E-Mail
  </label>
  <input
    data-testid="email-input"
    name="Email"
    onChange={[Function]}
    placeholder="Email"
    type="text"
    value=""
  />
  <label
    htmlFor="Message"
  >
    Message
  </label>
  <textarea
    data-testid="message-input"
    name="Message"
    onChange={[Function]}
    placeholder="Message"
    rows={5}
    value=""
  />
  <button
    data-testid="button"
    type="submit"
  >
    Submit
  </button>
</form>
`;
